File: ruby/logbrew-ruby/lib/logbrew/product_timeline.rb
# frozen_string_literal: true

module LogBrew
  # Builders for app-owned product and network timeline action events.
  class ProductTimeline
    private_class_method :new

    def self.product_action(
      name:,
      status: "success",
      route_template: nil,
      session_id: nil,
      trace_id: nil,
      screen: nil,
      funnel: nil,
      step: nil,
      metadata: nil
    )
      action_metadata = timeline_metadata("product_timeline", metadata)
      put_if_present(action_metadata, "routeTemplate", sanitize_optional_route_template("product route_template", route_template))
      put_if_present(action_metadata, "sessionId", optional_label("session_id", session_id))
      put_if_present(action_metadata, "traceId", optional_label("trace_id", trace_id))
      put_if_present(action_metadata, "screen", optional_label("screen", screen))
      put_if_present(action_metadata, "funnel", optional_label("funnel", funnel))
      put_if_present(action_metadata, "step", optional_label("step", step))

      {
        "name" => required_label("product action name", name),
        "status" => normalize_status(status),
        "metadata" => action_metadata
      }
    end

    def self.network_milestone(
      route_template:,
      method: "GET",
      status_code: nil,
      duration_ms: nil,
      status: nil,
      name: nil,
      session_id: nil,
      trace_id: nil,
      metadata: nil
    )
      route = sanitize_route_template("network milestone route_template", route_template)
      normalized_method = normalize_method(method)
      validate_status_code(status_code)
      validate_duration_ms(duration_ms)
      normalized_status = normalize_status(status || (status_code && status_code >= 400 ? "failure" : "success"))

      action_metadata = timeline_metadata("network_timeline", metadata)
      action_metadata["routeTemplate"] = route
      action_metadata["method"] = normalized_method
      put_if_present(action_metadata, "statusCode", status_code)
      put_if_present(action_metadata, "durationMs", duration_ms)
      put_if_present(action_metadata, "sessionId", optional_label("session_id", session_id))
      put_if_present(action_metadata, "traceId", optional_label("trace_id", trace_id))

      {
        "name" => name.nil? ? "network.#{normalized_method.downcase} #{route}" : required_label("network milestone name", name),
        "status" => normalized_status,
        "metadata" => action_metadata
      }
    end

    def self.timeline_metadata(source, metadata)
      copied = { "source" => source }
      validated = Validation.require_metadata(metadata)
      return copied if validated.nil?

      validated.each do |key, value|
        copied[key] = value unless key == "source"
      end
      copied
    end

    def self.required_label(label, value)
      Validation.require_non_empty(label, value)
      value.to_s.strip
    end

    def self.optional_label(label, value)
      return nil if value.nil?

      required_label(label, value)
    end

    def self.normalize_status(status)
      Validation.require_allowed_value("action status", status, LogBrew::ACTION_STATUSES)
      status
    end

    def self.sanitize_optional_route_template(label, route_template)
      return nil if route_template.nil?

      sanitize_route_template(label, route_template)
    end

    def self.sanitize_route_template(label, route_template)
      Validation.require_non_empty(label, route_template)
      trimmed = route_template.to_s.strip
      if trimmed.match?(/\Ahttps?:\/\//i)
        uri = URI.parse(trimmed)
        return uri.path.nil? || uri.path.empty? ? "/" : uri.path
      end

      cutoff = first_present_index(trimmed.index("?"), trimmed.index("#"))
      cutoff.nil? ? trimmed : (trimmed[0...cutoff].rstrip.empty? ? "/" : trimmed[0...cutoff].rstrip)
    rescue URI::InvalidURIError => error
      raise SdkError.new("validation_error", "route_template must be a valid route or URL: #{error.message}")
    end

    def self.normalize_method(method)
      normalized = method.to_s.strip.upcase
      unless normalized.match?(/\A[A-Z0-9_-]+\z/)
        raise SdkError.new("validation_error", "network milestone method must be a valid HTTP method")
      end

      normalized
    end

    def self.validate_status_code(status_code)
      return if status_code.nil?

      unless status_code.is_a?(Integer) && status_code >= 100 && status_code <= 599
        raise SdkError.new("validation_error", "network milestone status_code must be between 100 and 599")
      end
    end

    def self.validate_duration_ms(duration_ms)
      return if duration_ms.nil?
      unless duration_ms.is_a?(Numeric) && duration_ms.finite?
        raise SdkError.new("validation_error", "network milestone duration_ms must be a finite number")
      end
      raise SdkError.new("validation_error", "network milestone duration_ms must be non-negative") if duration_ms.negative?
    end

    def self.put_if_present(metadata, key, value)
      metadata[key] = value unless value.nil?
    end

    def self.first_present_index(first, second)
      return second if first.nil?
      return first if second.nil?

      [first, second].min
    end

    private_class_method :timeline_metadata, :required_label, :optional_label, :normalize_status,
                         :sanitize_optional_route_template, :sanitize_route_template, :normalize_method,
                         :validate_status_code, :validate_duration_ms, :put_if_present, :first_present_index
  end
end

File: php/logbrew-php/src/ProductTimeline.php
<?php

declare(strict_types=1);

namespace LogBrew;

/**
 * App-owned product and network timeline helpers.
 *
 * @phpstan-import-type ActionAttributes from LogBrewClient
 * @phpstan-import-type Metadata from LogBrewClient
 */
final class ProductTimeline
{
    /** @var list<string> */
    private const ACTION_STATUSES = ['queued', 'running', 'success', 'failure'];

    private function __construct()
    {
    }

    /**
     * Create an action attribute payload for a product step already known by the application.
     *
     * @param array<string, mixed> $metadata
     * @return ActionAttributes
     */
    public static function productAction(
        string $name,
        string $status = 'success',
        ?string $routeTemplate = null,
        ?string $sessionId = null,
        ?string $traceId = null,
        ?string $screen = null,
        ?string $funnel = null,
        ?string $step = null,
        array $metadata = []
    ): array {
        $normalizedStatus = self::normalizeStatus($status);
        $actionMetadata = self::createMetadata('product_timeline', $metadata);
        self::putIfNotNull($actionMetadata, 'routeTemplate', self::sanitizeOptionalRouteTemplate('product routeTemplate', $routeTemplate));
        self::putIfNotNull($actionMetadata, 'sessionId', self::label('sessionId', $sessionId));
        self::putIfNotNull($actionMetadata, 'traceId', self::label('traceId', $traceId));
        self::putIfNotNull($actionMetadata, 'screen', self::label('screen', $screen));
        self::putIfNotNull($actionMetadata, 'funnel', self::label('funnel', $funnel));
        self::putIfNotNull($actionMetadata, 'step', self::label('step', $step));

        return [
            'name' => self::requiredLabel('product action name', $name),
            'status' => $normalizedStatus,
            'metadata' => $actionMetadata,
        ];
    }

    /**
     * Create an action attribute payload for an app-owned API or network milestone.
     *
     * @param array<string, mixed> $metadata
     * @return ActionAttributes
     */
    public static function networkMilestone(
        string $routeTemplate,
        string $method = 'GET',
        ?int $statusCode = null,
        int|float|null $durationMs = null,
        ?string $status = null,
        ?string $name = null,
        ?string $sessionId = null,
        ?string $traceId = null,
        array $metadata = []
    ): array {
        $route = self::sanitizeRouteTemplate('network milestone routeTemplate', $routeTemplate);
        $normalizedMethod = self::normalizeMethod($method);
        self::validateStatusCode($statusCode);
        self::validateDurationMs($durationMs);
        $normalizedStatus = self::normalizeStatus($status ?? ($statusCode !== null && $statusCode >= 400 ? 'failure' : 'success'));

        $actionMetadata = self::createMetadata('network_timeline', $metadata);
        $actionMetadata['routeTemplate'] = $route;
        $actionMetadata['method'] = $normalizedMethod;
        self::putIfNotNull($actionMetadata, 'statusCode', $statusCode);
        self::putIfNotNull($actionMetadata, 'durationMs', $durationMs);
        self::putIfNotNull($actionMetadata, 'sessionId', self::label('sessionId', $sessionId));
        self::putIfNotNull($actionMetadata, 'traceId', self::label('traceId', $traceId));

        return [
            'name' => $name === null ? sprintf('network.%s %s', strtolower($normalizedMethod), $route) : self::requiredLabel('network milestone name', $name),
            'status' => $normalizedStatus,
            'metadata' => $actionMetadata,
        ];
    }

    private static function label(string $label, ?string $value): ?string
    {
        if ($value === null) {
            return null;
        }

        return self::requiredLabel($label, $value);
    }

    private static function requiredLabel(string $label, string $value): string
    {
        LogBrewClient::requireNonEmpty($label, $value);
        return trim($value);
    }

    /** @return 'queued'|'running'|'success'|'failure' */
    private static function normalizeStatus(string $status): string
    {
        LogBrewClient::requireNonEmpty('action status', $status);
        return match ($status) {
            'queued', 'running', 'success', 'failure' => $status,
            default => throw new SdkError('validation_error', 'action status must be one of: ' . implode(', ', self::ACTION_STATUSES)),
        };
    }

    private static function sanitizeOptionalRouteTemplate(string $label, ?string $routeTemplate): ?string
    {
        if ($routeTemplate === null) {
            return null;
        }

        return self::sanitizeRouteTemplate($label, $routeTemplate);
    }

    private static function sanitizeRouteTemplate(string $label, string $routeTemplate): string
    {
        LogBrewClient::requireNonEmpty($label, $routeTemplate);
        $trimmed = trim($routeTemplate);
        $parts = parse_url($trimmed);
        if (is_array($parts) && isset($parts['scheme'], $parts['host'])) {
            $path = (string) ($parts['path'] ?? '/');
            return $path === '' ? '/' : $path;
        }

        $cutoff = self::firstPresentIndex(strpos($trimmed, '?'), strpos($trimmed, '#'));
        if ($cutoff !== null) {
            $trimmed = rtrim(substr($trimmed, 0, $cutoff));
        }

        return $trimmed === '' ? '/' : $trimmed;
    }

    private static function normalizeMethod(string $method): string
    {
        $normalized = strtoupper(trim($method));
        if ($normalized === '' || preg_match('/^[A-Z0-9_-]+$/', $normalized) !== 1) {
            throw new SdkError('validation_error', 'network milestone method must be a valid HTTP method');
        }

        return $normalized;
    }

    private static function validateStatusCode(?int $statusCode): void
    {
        if ($statusCode !== null && ($statusCode < 100 || $statusCode > 599)) {
            throw new SdkError('validation_error', 'network milestone statusCode must be between 100 and 599');
        }
    }

    private static function validateDurationMs(int|float|null $durationMs): void
    {
        if ($durationMs === null) {
            return;
        }
        if (is_float($durationMs) && !is_finite($durationMs)) {
            throw new SdkError('validation_error', 'network milestone durationMs must be a finite number');
        }
        if ($durationMs < 0) {
            throw new SdkError('validation_error', 'network milestone durationMs must be non-negative');
        }
    }

    /**
     * @param array<string, mixed> $metadata
     * @return Metadata
     */
    private static function createMetadata(string $source, array $metadata): array
    {
        $copied = ['source' => $source];
        foreach ($metadata as $key => $value) {
            $stringKey = (string) $key;
            LogBrewClient::requireNonEmpty('metadata key', $stringKey);
            $metadataValue = self::metadataValue($stringKey, $value);
            if ($stringKey !== 'source') {
                $copied[$stringKey] = $metadataValue;
            }
        }

        return $copied;
    }

    private static function metadataValue(string $key, mixed $value): string|int|float|bool|null
    {
        if ($value === null || is_string($value) || is_int($value) || is_bool($value)) {
            return $value;
        }
        if (is_float($value) && is_finite($value)) {
            return $value;
        }

        throw new SdkError('validation_error', sprintf('metadata value for %s must be a string, number, boolean, or null', $key));
    }

    /**
     * @param Metadata $metadata
     */
    private static function putIfNotNull(array &$metadata, string $key, string|int|float|bool|null $value): void
    {
        if ($value !== null) {
            $metadata[$key] = $value;
        }
    }

    private static function firstPresentIndex(int|false $first, int|false $second): ?int
    {
        if ($first === false) {
            return $second === false ? null : $second;
        }
        if ($second === false) {
            return $first;
        }

        return min($first, $second);
    }
}
